google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -515,70 +843,173 @@ private void verifyNotClosed() {
    */
-  private BulkCommitBatch getEligibleBatch() {
+  private BulkCommitBatch getEligibleBatch(
+      DocumentReference documentReference, List<BulkCommitBatch> batchQueue) {
     if (batchQueue.size() > 0) {
       BulkCommitBatch lastBatch = batchQueue.get(batchQueue.size() - 1);
-      if (lastBatch.getState() == UpdateBuilder.BatchState.OPEN) {
+      if (lastBatch.isOpen() && !lastBatch.has(documentReference)) {
         return lastBatch;
       }
     }
-    return createNewBatch();
+    return createNewBatch(batchQueue);
   }
 
   /**
    * Creates a new batch and adds it to the BatchQueue. If there is already a batch enqueued, sends
    * the batch after a new one is created.
    */
-  private BulkCommitBatch createNewBatch() {
+  private BulkCommitBatch createNewBatch(List<BulkCommitBatch> batchQueue) {
     BulkCommitBatch newBatch = new BulkCommitBatch(firestore, maxBatchSize);
 
     if (batchQueue.size() > 0) {
       batchQueue.get(batchQueue.size() - 1).markReadyToSend();
-      sendReadyBatches();
     }
     batchQueue.add(newBatch);
     return newBatch;
   }
 
   /**
-   * Attempts to send batches starting from the front of the BatchQueue until a batch cannot be
-   * sent.
+   * Attempts to send batches starting from the front of the provided batch queue until a batch
+   * cannot be sent.
    *
    * <p>After a batch is complete, try sending batches again.
    */
-  private void sendReadyBatches() {
-    List<BulkCommitBatch> unsentBatches =
-        FluentIterable.from(batchQueue)
-            .filter(
-                new Predicate<BulkCommitBatch>() {
-                  @Override
-                  public boolean apply(BulkCommitBatch batch) {
-                    return batch.getState() == UpdateBuilder.BatchState.READY_TO_SEND;
-                  }
-                })
-            .toList();
-
+  private void sendReadyBatches(final List<BulkCommitBatch> batchQueue) {
     int index = 0;
-    while (index < unsentBatches.size()
-        && unsentBatches.get(index).state == BatchState.READY_TO_SEND) {
-      final BulkCommitBatch batch = unsentBatches.get(index);
+    while (index < batchQueue.size() && batchQueue.get(index).isReadyToSend()) {
+      final BulkCommitBatch batch = batchQueue.get(index);
+
+      // Future that completes when the current batch or its scheduling attempts completes.
+      final SettableApiFuture<Void> batchCompletedFuture = SettableApiFuture.create();
+      pendingBatches.add(batchCompletedFuture);
 
       // Send the batch if it is under the rate limit, or schedule another attempt after the
       // appropriate timeout.
       long delayMs = rateLimiter.getNextRequestDelayMs(batch.getPendingOperationCount());
       Preconditions.checkState(delayMs != -1, "Batch size should be under capacity");
       if (delayMs == 0) {
-        sendBatch(batch);
+        sendBatch(batch, batchQueue, batchCompletedFuture);
       } else {
         firestoreExecutor.schedule(
```

Comment:
Can we use our new executor here as well? It would need to be a `ScheduledExecutor`.